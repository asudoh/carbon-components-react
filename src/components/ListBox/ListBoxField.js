import React from 'react';
import PropTypes from 'prop-types';
import ListBoxMenuIcon from './ListBoxMenuIcon';
import ListBoxSelection from './ListBoxSelection';
import childrenOf from '../../prop-types/childrenOf';

/**
 * `ListBoxField` is responsible for creating the containing node for valid
 * elements inside of a field. It also provides a11y-related attributes like
 * `role` to make sure a user can focus the given field.
 */
const ListBoxField = ({ children, prefix, ...rest }) => (
  <div
    role="button"
    className={`${prefix}--list-box__field`}
    tabIndex="0"
    {...rest}>
    {children}
  </div>
);

ListBoxField.propTypes = {
  /**
   * Provide the contents of your ListBoxField
   */
  children: childrenOf([ListBoxMenuIcon, ListBoxSelection, 'span', 'input']),

  /**
   * The selector prefix
   */
  prefix: PropTypes.string,
};

ListBoxField.defaultProps = {
  prefix: 'bx',
};

export default ListBoxField;
