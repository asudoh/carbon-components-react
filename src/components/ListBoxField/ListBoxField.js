import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import ListBoxLabel from '../ListBoxLabel';
import ListBoxMenuIcon from '../ListBoxMenuIcon';

const ListBoxField = ({
  type = 'default',
  label,
  selectedItem,
  clearSelection,
  isOpen,
  // `rest` here should correspond with `getButtonProps` from `downshift`
  ...rest,
}) => {
  const className = cx({
    'bx--list-box__field': true,
    'bx--list-box__field--inline': type === 'inline',
    'bx--list-box__field--open': isOpen,
  });
  return (
    <div className={className} tabIndex="0" {...rest}>
      <ListBoxLabel
        label={label}
        count={selectedItem.length}
        clearSelection={clearSelection}
      />
      <ListBoxMenuIcon isOpen={isOpen} />
    </div>
  );
};

ListBoxField.propTypes = {
  type: PropTypes.oneOf(['default', 'inline']),
  label: PropTypes.string.isRequired,
  selectedItem: PropTypes.array.isRequired,
  clearSelection: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
};

export default ListBoxField;
