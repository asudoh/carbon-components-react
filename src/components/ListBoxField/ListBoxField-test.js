import React from 'react';
import { shallow } from 'enzyme';
import ListBoxField from '../ListBoxField';
import { generateItems, generateGenericItem } from '../../tools/testing/items';

describe('ListBoxField', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      selectedItem: generateItems(5, generateGenericItem),
      clearSelection: jest.fn(),
    };
  });

  it('should render', () => {
    const wrapper = shallow(
      <ListBoxField label="Label" isOpen={false} {...mockProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should be focus-able', () => {
    const wrapper = shallow(
      <ListBoxField label="Label" isOpen={false} {...mockProps} />
    );
    expect(wrapper.prop('tabIndex')).toBe('0');
  });

  it('should add an inline class if `type="inline"` is provided', () => {
    const wrapper = shallow(
      <ListBoxField type="inline" label="Label" isOpen={false} {...mockProps} />
    );
    expect(wrapper.find('.bx--list-box__field--inline').length).toBe(1);
  });
});
