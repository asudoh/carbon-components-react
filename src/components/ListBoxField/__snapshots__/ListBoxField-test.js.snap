// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBoxField should render 1`] = `
<div
  className="bx--list-box__field"
  tabIndex="0"
>
  <ListBoxLabel
    clearSelection={[Function]}
    count={5}
    label="Label"
  />
  <ListBoxMenuIcon
    isOpen={false}
  />
</div>
`;
